File: harness/bunnymark/benchmarks/BunnymarkComputation/cs/BunnymarkComputation.cs
using System;
using System.Collections.Generic;
using Godot;

public partial class BunnymarkComputation : Node2D
{
    private sealed class BunnyData
    {
        public Vector2 Position;
        public Vector2 Center;
        public double Phase;
        public double Frequency;
        public Vector2 Radius;

        public BunnyData(Vector2 position, Vector2 center, double phase, double frequency, Vector2 radius)
        {
            Position = position;
            Center = center;
            Phase = phase;
            Frequency = frequency;
            Radius = radius;
        }
    }

    private readonly List<BunnyData> _bunnies = new();
    private readonly Texture2D _bunnyTexture = ResourceLoader.Load<Texture2D>("res://images/godot_bunny.png");
    private readonly RandomNumberGenerator _randomNumberGenerator = new();

    private Vector2 _screenSize;

    public override void _Ready()
    {
        _randomNumberGenerator.Randomize();
    }

    public override void _Draw()
    {
        foreach (BunnyData bunny in _bunnies)
        {
            DrawTexture(_bunnyTexture, bunny.Position);
        }
    }

    public override void _Process(double delta)
    {
        _screenSize = GetViewportRect().Size;

        foreach (BunnyData bunny in _bunnies)
        {
            bunny.Phase += bunny.Frequency * delta;
            double phase = bunny.Phase;

            double offsetX = Math.Sin(phase) * bunny.Radius.X;
            double offsetY = Math.Cos(phase * 1.7) * bunny.Radius.Y;
            double distance = Math.Sqrt(offsetX * offsetX + offsetY * offsetY);
            double angle = Math.Atan2(offsetY, offsetX) + Math.Sin(phase * 0.5) * 0.75;

            Vector2 pos = bunny.Position;
            pos.X = (float)(bunny.Center.X + Math.Cos(angle) * distance);
            pos.Y = (float)(bunny.Center.Y + Math.Sin(angle) * distance);
            bunny.Position = pos;
        }
        QueueRedraw();
    }

    public void add_bunny()
    {
        _bunnies.Add(new BunnyData(
            new Vector2(0, 0),
            new Vector2(
                _randomNumberGenerator.Randf() * _screenSize.X,
                _randomNumberGenerator.Randf() * _screenSize.Y
            ),
            _randomNumberGenerator.Randf() * 6.283185307179586,
            _randomNumberGenerator.Randf() * 2.0 + 0.5,
            new Vector2(
                _randomNumberGenerator.Randf() * _screenSize.X / 8,
                _randomNumberGenerator.Randf() * _screenSize.Y / 8
            )
        ));
    }

    public void remove_bunny()
    {
        if (_bunnies.Count == 0) return;
        _bunnies.RemoveAt(_bunnies.Count - 1);
    }

    public void finish()
    {
        EmitSignal("benchmark_finished", _bunnies.Count);
    }
}

File: harness/bunnymark/src/main/kotlin/godot/benchmark/bunnymark/BunnymarkComputation.kt
package godot.benchmark.bunnymark

import godot.api.Node2D
import godot.api.RandomNumberGenerator
import godot.api.ResourceLoader
import godot.api.Texture2D
import godot.annotation.Script
import godot.annotation.Register
import godot.annotation.Emit
import godot.core.Vector2
import godot.core.signal1
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

@Script("BunnymarkComputation")
class BunnymarkComputation : Node2D() {

	@Emit
	val benchmarkFinished by signal1<Int>()

	data class Bunny(
		var position: Vector2,
		var center: Vector2,
		var phase: Double,
		var frequency: Double,
		var radius: Vector2,
	)

	private val bunnies = mutableListOf<Bunny>()
	private val bunnyTexture = ResourceLoader.load("res://images/godot_bunny.png") as Texture2D
	private val randomNumberGenerator = RandomNumberGenerator()

	private lateinit var screenSize: Vector2

	@Register
	override fun _ready() {
		randomNumberGenerator.randomize()
	}

	@Register
	override fun _draw() {
		for (bunny in bunnies) {
			drawTexture(bunnyTexture, bunny.position)
		}
	}

	@Register
	override fun _process(delta: Double) {
		screenSize = getViewportRect().size

		for (bunny in bunnies) {
			bunny.phase += bunny.frequency * delta
			val phase = bunny.phase

			val offsetX = sin(phase) * bunny.radius.x
			val offsetY = cos(phase * 1.7) * bunny.radius.y
			val distance = sqrt(offsetX * offsetX + offsetY * offsetY)
			val angle = atan2(offsetY, offsetX) + sin(phase * 0.5) * 0.75

			val pos = bunny.position
			pos.x = bunny.center.x + cos(angle) * distance
			pos.y = bunny.center.y + sin(angle) * distance
			bunny.position = pos
		}
		queueRedraw()
	}

	@Register
	fun addBunny() {
		bunnies.add(
			Bunny(
				Vector2(0, 0),
				Vector2(
					randomNumberGenerator.randf().toDouble() * screenSize.x,
					randomNumberGenerator.randf().toDouble() * screenSize.y
				),
				randomNumberGenerator.randf().toDouble() * 6.283185307179586,
				randomNumberGenerator.randf().toDouble() * 2.0 + 0.5,
				Vector2(
					randomNumberGenerator.randf().toDouble() * screenSize.x / 8,
					randomNumberGenerator.randf().toDouble() * screenSize.y / 8
				)
			)
		)
	}

	@Register
	fun removeBunny() {
		if (bunnies.isEmpty()) return
		bunnies.removeAt(bunnies.size - 1)
	}

	@Register
	fun finish() {
		benchmarkFinished.emit(bunnies.size)
	}
}
